### PR TITLE
Allow to define a custom ImagePullPolicy via configuration

### LINKED ISSUE #7405: [Enhancement]: Add global configuration for pull policy
### Module

Core

### Proposal

Hi,

Currently it is necessary to configure the image pull policy for each and every container. It would be very useful if this could be configured globally, e.g. via the _testcontainers.properties_ file. 

Thank you and best regards

  Nils

### PR DESCRIPTION
**Pull request by eddumelendez**
Fixes #7405

### CODE REVIEW

File: core/src/test/java/org/testcontainers/images/OverrideImagePullPolicyTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            .when(TestcontainersConfiguration.getInstance())
+            .getImagePullPolicy();
+
+        try (DockerRegistryContainer registry = new DockerRegistryContainer()) {
+            registry.start();
+            GenericContainer<?> container = new GenericContainer<>(registry.createImage()).withExposedPorts(8080);
+            container.start();
+            assertThat(container.getImage())
+                .asString()
+                .contains("imagePullPolicy=org.testcontainers.utility.FakeImagePullPolicy");
```
Comment (kiview) [inline core/src/test/java/org/testcontainers/images/OverrideImagePullPolicyTest.java:52]: I don't really like the assert of this test. We are asserting against an unrelated implementation detail here (`toString()` of `RemoteDockerImage`).

Two ideas how to improve the test:
* Either make `imagePullPolicy` visible for testing and assert against the field.
* Try to achieve a test similar to `AgeBasedPullPolicyTest` or `ImagePullPolicyTest`, where we check that the expected pull policy is used, by asserting that an image is pulled.

WDYT?

**Review by kiview — CHANGES_REQUESTED** [review]
I think we should have a better test before merging (but open to discuss it), see comments.

File: core/src/main/java/org/testcontainers/images/PullPolicy.java [hunk 2]
```diff
@@ -1,22 +1,60 @@
...
+                    configuredInstance =
+                        (ImagePullPolicy) Thread
+                            .currentThread()
+                            .getContextClassLoader()
+                            .loadClass(imagePullPolicyClassName)
+                            .getConstructor()
+                            .newInstance();
+                } catch (Exception e) {
+                    throw new IllegalArgumentException(
+                        "Configured Pull Policy could not be loaded: " + imagePullPolicyClassName,
```
[the remark this answers is not in the document]
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/images/PullPolicy.java:44]: Thanks for the note. I think is much better to align those as they are referring to the implementation class for `ImagePullPolicy`

File: core/src/test/java/org/testcontainers/images/OverrideImagePullPolicyTest.java (hunk 1 above)
Comment (eddumelendez) [reply core/src/test/java/org/testcontainers/images/OverrideImagePullPolicyTest.java:52]: > Either make imagePullPolicy visible for testing and assert against the field.

Given we are testing the right `imagePullPolicy` has been applied, this sounds like the one to do.

**Review by kiview — APPROVED** [review]

File: core/src/main/java/org/testcontainers/images/PullPolicy.java [hunk 3]
```diff
@@ -1,22 +1,62 @@
...
+    @VisibleForTesting
+    static ImagePullPolicy defaultImplementation = new DefaultPullPolicy();
+
     /**
      * Convenience method for returning the {@link DefaultPullPolicy} default image pull policy
      * @return {@link ImagePullPolicy}
      */
     public static ImagePullPolicy defaultPolicy() {
-        return new DefaultPullPolicy();
+        if (instance == null) {
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/images/PullPolicy.java:29]: 1. Are we okay with double instantiation when it is used concurrently? Or should we mark it as `synchronized`?
2. minor nit: consider doing an early return on `instance != null` to reduce the indentation 👍

**Review by bsideup — DISMISSED** [review]

File: core/src/main/java/org/testcontainers/images/PullPolicy.java (hunk 3 above)
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/images/PullPolicy.java]: thanks! I've applied changes and add docs too.

**Comment by OLibutzki** [comment]
Thanks for providing this helpful configuration!

@eddumelendez ,
does the configuration affect the pull policy for database containers launched via JDBC URL scheme, too?

**Comment by eddumelendez** [comment]
Yes, it should. Testcontainers JDBC URL uses the Database container implementation under the hood.

**Pull request merged**